File: common/src/main/java/xyz/chunkstories/net/packets/PacketSendFile.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.net.packets;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

import xyz.chunkstories.api.exceptions.PacketProcessingException;
import xyz.chunkstories.api.net.Packet;
import xyz.chunkstories.api.net.PacketDestinator;
import xyz.chunkstories.api.net.PacketReceptionContext;
import xyz.chunkstories.api.net.PacketSender;
import xyz.chunkstories.api.net.PacketSendingContext;

public class PacketSendFile extends Packet {
	public String fileTag;
	public File file;

	@Override
	public void send(PacketDestinator destinator, DataOutputStream out, PacketSendingContext ctx) throws IOException {
		out.writeUTF(fileTag);

		if (file.exists()) {
			out.writeLong(file.length());
			FileInputStream fis = new FileInputStream(file);
			byte[] buffer = new byte[4096];
			int read;
			while (true) {
				read = fis.read(buffer);
				// System.out.println("read"+read);
				if (read > 0)
					out.write(buffer, 0, read);
				else
					break;
			}
			fis.close();
		} else
			out.writeLong(0L);
	}

	@Override
	public void process(PacketSender sender, DataInputStream in, PacketReceptionContext processor)
			throws IOException, PacketProcessingException {
		// Ignore packets incomming
	}
}

File: common/src/main/java/xyz/chunkstories/net/packets/PacketSendFile.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.net.packets

import xyz.chunkstories.api.net.Packet
import xyz.chunkstories.api.player.Player
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream

open class PacketSendFile : Packet() {
    lateinit var fileTag: String
    lateinit var file: File

    override fun send(dos: DataOutputStream) {
        dos.writeUTF(fileTag)
        if (file.exists()) {
            dos.writeLong(file.length())
            val fis = FileInputStream(file)
            val buffer = ByteArray(4096)
            var read: Int
            while (true) {
                read = fis.read(buffer)
                if (read > 0) dos.write(buffer, 0, read) else break
            }
            fis.close()
        } else dos.writeLong(0L)
    }

    override fun receive(dis: DataInputStream, player: Player?) {
        throw Exception("Unexpected")
    }
}